REST API for core

tdudzik opened the pull request:
REST API for core.

Possibilities:
- create list of books: POST http://localhost:8090
- retrieve all books: GET http://localhost:8090

I've wrote also a mapper (DTO -> Entity, Entity -> DTO) using the mapper from http://modelmapper.org/.

Commit:
KoboBooks parser
I've create prototype parser for kobobooks.com. Parser get almost all data about books that are required (except tags)
This implementation need few refactors to be done and more optimalization, bacause parsing from that site have high time complexity.

Commit:
Refactor of logic
I've fixed and extended logic of parser for kobobooks.com page. I still need to get rid of msvm in KoboBooksParser but I'll do this as soon as I rebese master on this branch. Then I'll add some tests for it. I know I broke TDD manner but in this case developing in TDD manner will be to hard and time consuming. Most of a time I spend on debugging and finding out how can I get date from html.

Commit:
Rebased master on branch and little refactor
I've rebased maseter onto this branch, resolved conflicts and did little refactor.

Commit:
Fix previous commits
By mistake I did merged branches properly. This commit fixes it

Commit:
Refactor after code review
I've refactored all classes and connected them to common 'downloader'. It's first step into ingegration with rest of parsers.
Also I've added looging to all classes - 'log them all'

tdudzik committed:
Add required dependencies to pom.xml.

tdudzik committed:
Add configuration of Spring Boot and Hibernate.

tdudzik committed:
Create entities and repositories.

tdudzik committed:
Create data transfer object that will be used in rest controllers.

tdudzik committed:
Create mappers from DTO to entity and from entity to DTO.

tdudzik committed:
Create unit tests for Library entity in order to test if Library is properly created from string.

tdudzik committed:
Create rest controller for books. It is possible to create a list of books or get all books from database.

Commit:
Second step in integrating components
I've refactored once again whole kobobooks parser package. Now it has more things in common with others parsers

Commit:
Restoring previous code
By mistake I've deleted files that should not be deleted and commited those changes

caroo7 commented on Common/src/main/java/javapocalypse/common/dto/BookDTO.java:
```diff
@@ -0,0 +1,57 @@
+
+    @Getter
+    @Setter
+    private long id;
```
Why you need 'id' in DTO?

Commit:
Code format and optimize import
I went through all classes and I've used format and optimize imports on them. Nothing fancy :smile:

caroo7 commented on Core/src/main/java/javapocalypse/core/api/BooksController.java:
```diff
@@ -0,0 +1,42 @@
+        return bookMapper.mapToDTOs(bookRepository.findAll());
+    }
+
+    @RequestMapping(path = "", method = RequestMethod.POST, consumes = "application/json", produces = "application/json")
```
:+1:

egolesor replied:
@PostMapping ?

tdudzik replied:
@egolesor  what do you mean?

caroo7 commented:
:+1:

egolesor commented on Common/src/main/java/javapocalypse/common/dto/BookDTO.java:
```diff
@@ -0,0 +1,57 @@
+                .toString();
+    }
+
+}
```
Where you need all this setters and getters ?

egolesor commented on Core/src/main/java/javapocalypse/core/Application.java:
```diff
@@ -0,0 +1,16 @@
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.annotation.ComponentScan;
+
+@ComponentScan
```
I think ComponentScan is not needed and the variable ctx albo is not used so you can just write SpringApplication.run(Application.class). Maybe you have a reason for that ?

egolesor commented on Core/src/main/java/javapocalypse/core/api/BooksController.java:
```diff
@@ -0,0 +1,42 @@
+import java.util.List;
+
+@RestController
+@EnableAutoConfiguration
```
Why You enable this auto configuration in the controller  ?

egolesor commented on Core/src/main/java/javapocalypse/core/api/BooksController.java:
```diff
@@ -0,0 +1,42 @@
+    private final BookMapper bookMapper;
+
+    @Inject
+    public BooksController(BookRepository bookRepository, BookMapper bookMapper) {
```
:+1:

egolesor commented on Core/src/main/java/javapocalypse/core/api/BooksController.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @RequestMapping(path = "", method = RequestMethod.GET, produces = "application/json")
+    @ResponseBody
```
@RestController automatically add @ResponseBody at all methods so you may delete this.

egolesor commented on Core/src/main/java/javapocalypse/core/api/BooksController.java:
```diff
@@ -0,0 +1,42 @@
+    }
+
+    @RequestMapping(path = "", method = RequestMethod.POST, consumes = "application/json", produces = "application/json")
+    @ResponseBody
```
Here again is no need for that

egolesor commented on Core/src/main/java/javapocalypse/core/model/entities/Book.java:
```diff
@@ -0,0 +1,77 @@
+    @Column(name = "id")
+    private long id;
+
+    @Enumerated(EnumType.STRING)
```
:+1:

egolesor commented on Core/src/main/java/javapocalypse/core/model/entities/Book.java:
```diff
@@ -0,0 +1,77 @@
+    public Book() {
+    }
+
+    public Book(
```
For this 2 methods you can use lombok if you used it so far

tdudzik replied:
No I can't. Notice that the second constructor does not have `id` param.

egolesor commented on Core/src/main/java/javapocalypse/core/model/entities/Library.java:
```diff
@@ -0,0 +1,23 @@
+    EMPIK_COM {
+        @Override
+        public String toString() {
+            return "empik_com";
```
first time I see such implementation :+1:

egolesor commented:
After fixing the issues change the wiki and the script to run the application or inform me to do that.

Commit:
Fixes after code review
I've changed few things after code review

tdudzik committed:
Fixes after a code review.

tdudzik commented:
Me and Łukasz fucked up my branch and I need to create another pull request after resolving issues with rebasing.
